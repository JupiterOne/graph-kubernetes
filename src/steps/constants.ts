import {
  RelationshipClass,
  RelationshipDirection,
  StepMappedRelationshipMetadata,
} from '@jupiterone/integration-sdk-core';

export const CLUSTER_ENTITY_DATA_KEY = 'entity:cluster';

export enum IntegrationSteps {
  NETWORK_POLICIES = 'fetch-network-policies',
  SERVICE_ACCOUNTS = 'fetch-service-accounts',
  USERS = 'fetch-users',
  ROLES = 'fetch-roles',
  CLUSTER_ROLES = 'fetch-cluster-roles',
  ROLE_BINDINGS = 'fetch-role-bindings',
  CLUSTER_ROLE_BINDINGS = 'fetch-cluster-role-bindings',
  CLUSTERS = 'fetch-clusters',
  BUILD_CLUSTER_RESOURCES_RELATIONSHIPS = 'build-cluster-resources-relationships',
  BUILD_CLUSTER_AKS_RELATIONSHIPS = 'build-cluster-aks-relationships',
  BUILD_CLUSTER_GKE_RELATIONSHIPS = 'build-cluster-gke-relationships',
  NAMESPACES = 'fetch-namespaces',
  NODES = 'fetch-nodes',
  SERVICES = 'fetch-services',
  DEPLOYMENTS = 'fetch-deployments',
  BUILD_CONTAINER_SPEC_VOLUME_RELATIONSHIPS = 'build-container-spec-volume-relationships',
  REPLICASETS = 'fetch-replica-sets',
  STATEFULSETS = 'fetch-stateful-sets',
  DAEMONSETS = 'fetch-daemon-sets',
  JOBS = 'fetch-jobs',
  CRONJOBS = 'fetch-cron-jobs',
  CONFIGMAPS = 'fetch-config-maps',
  SECRETS = 'fetch-secrets',
  IMAGES = 'fetch-images',
  PODS = 'fetch-pods',
}

export const Entities = {
  NETWORK_POLICY: {
    _type: 'kube_network_policy',
    _class: ['Configuration'],
    resourceName: 'Kubernetes Network Policy',
  },
  SERVICE_ACCOUNT: {
    _type: 'kube_service_account',
    _class: ['User'],
    resourceName: 'Kubernetes Service Account',
  },
  USER: {
    _type: 'kube_user',
    _class: ['User'],
    resourceName: 'Kubernetes User',
  },
  ROLE: {
    _type: 'kube_role',
    _class: ['AccessRole'],
    resourceName: 'Kubernetes Role',
  },
  CLUSTER_ROLE: {
    _type: 'kube_cluster_role',
    _class: ['AccessRole'],
    resourceName: 'Kubernetes Cluster Role',
  },
  ROLE_BINDING: {
    _type: 'kube_role_binding',
    _class: ['AccessPolicy'],
    resourceName: 'Kubernetes Role Binding',
  },
  CLUSTER_ROLE_BINDING: {
    _type: 'kube_cluster_role_binding',
    _class: ['AccessPolicy'],
    resourceName: 'Kubernetes Cluster Role Binding',
  },
  CLUSTER: {
    _type: 'kube_cluster',
    _class: ['Cluster'],
    resourceName: 'Kubernetes Cluster',
  },
  AZURE_KUBERNETES_CLUSTER: {
    _type: 'azure_kubernetes_cluster',
    _class: ['Cluster'],
    resourceName: 'Azure Kubernetes Cluster',
  },
  GOOGLE_KUBERNETES_CLUSTER: {
    _type: 'google_container_cluster',
    _class: ['Cluster'],
    resourceName: 'Google Kubernetes Cluster',
  },
  NAMESPACE: {
    _type: 'kube_namespace',
    _class: ['Group'],
    resourceName: 'Kubernetes Namespace',
  },
  DEPLOYMENT: {
    _type: 'kube_deployment',
    _class: ['Deployment'],
    resourceName: 'Kubernetes Deployment',
  },
  CONTAINER_SPEC: {
    _type: 'kube_container_spec',
    _class: ['Configuration'],
    resourceName: 'Kubernetes Container Spec',
  },
  VOLUME: {
    _type: 'kube_volume',
    _class: ['Disk'],
    resourceName: 'Kubernetes Volume',
  },
  REPLICASET: {
    _type: 'kube_replica_set',
    _class: ['Deployment'],
    resourceName: 'Kubernetes ReplicaSet',
  },
  STATEFULSET: {
    _type: 'kube_stateful_set',
    _class: ['Deployment'],
    resourceName: 'Kubernetes StatefulSet',
  },
  DAEMONSET: {
    _type: 'kube_daemon_set',
    _class: ['Deployment'],
    resourceName: 'Kubernetes DaemonSet',
  },
  NODE: {
    _type: 'kube_node',
    _class: ['Host'],
    resourceName: 'Kubernetes Node',
  },
  JOB: {
    _type: 'kube_job',
    _class: ['Task'],
    resourceName: 'Kubernetes Job',
  },
  CRONJOB: {
    _type: 'kube_cron_job',
    _class: ['Task'],
    resourceName: 'Kubernetes CronJob',
  },
  CONFIGMAP: {
    _type: 'kube_config_map',
    _class: ['Configuration'],
    resourceName: 'Kubernetes ConfigMap',
  },
  SECRET: {
    _type: 'kube_secret',
    _class: ['Vault'],
    resourceName: 'Kubernetes Secret',
  },
  SERVICE: {
    _type: 'kube_service',
    _class: ['Service'],
    resourceName: 'Kubernetes Service',
  },
  IMAGE: {
    _type: 'kube_image',
    _class: ['Image'],
    resourceName: 'Kubernetes Image',
  },
  POD: {
    _type: 'kube_pod',
    _class: ['Task'],
    resourceName: 'Kubernetes Pod',
  },
  CONTAINER: {
    _type: 'kube_container',
    _class: ['Container'],
    resourceName: 'Kubernetes Container',
  },
};

export const Relationships = {
  CLUSTER_CONTAINS_NAMESPACE: {
    _type: 'kube_cluster_contains_namespace',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.CLUSTER._type,
    targetType: Entities.NAMESPACE._type,
  },
  CLUSTER_CONTAINS_CLUSTER_ROLE: {
    _type: 'kube_cluster_contains_cluster_role',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.CLUSTER._type,
    targetType: Entities.CLUSTER_ROLE._type,
  },
  CLUSTER_CONTAINS_CLUSTER_ROLE_BINDING: {
    _type: 'kube_cluster_contains_cluster_role_binding',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.CLUSTER._type,
    targetType: Entities.CLUSTER_ROLE_BINDING._type,
  },
  NAMESPACE_CONTAINS_SERVICE: {
    _type: 'kube_namespace_contains_service',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.NAMESPACE._type,
    targetType: Entities.SERVICE._type,
  },
  NAMESPACE_CONTAINS_DEPLOYMENT: {
    _type: 'kube_namespace_contains_deployment',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.NAMESPACE._type,
    targetType: Entities.DEPLOYMENT._type,
  },
  NAMESPACE_CONTAINS_REPLICASET: {
    _type: 'kube_namespace_contains_replica_set',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.NAMESPACE._type,
    targetType: Entities.REPLICASET._type,
  },
  NAMESPACE_CONTAINS_STATEFULSET: {
    _type: 'kube_namespace_contains_stateful_set',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.NAMESPACE._type,
    targetType: Entities.STATEFULSET._type,
  },
  NAMESPACE_CONTAINS_DAEMONSET: {
    _type: 'kube_namespace_contains_daemon_set',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.NAMESPACE._type,
    targetType: Entities.DAEMONSET._type,
  },
  NAMESPACE_CONTAINS_CRONJOB: {
    _type: 'kube_namespace_contains_cron_job',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.NAMESPACE._type,
    targetType: Entities.CRONJOB._type,
  },
  NAMESPACE_CONTAINS_JOB: {
    _type: 'kube_namespace_contains_job',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.NAMESPACE._type,
    targetType: Entities.JOB._type,
  },
  NAMESPACE_CONTAINS_CONFIGMAP: {
    _type: 'kube_namespace_contains_config_map',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.NAMESPACE._type,
    targetType: Entities.CONFIGMAP._type,
  },
  NAMESPACE_CONTAINS_SECRET: {
    _type: 'kube_namespace_contains_secret',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.NAMESPACE._type,
    targetType: Entities.SECRET._type,
  },
  NAMESPACE_CONTAINS_NETWORK_POLICY: {
    _type: 'kube_namespace_contains_network_policy',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.NAMESPACE._type,
    targetType: Entities.NETWORK_POLICY._type,
  },
  NAMESPACE_CONTAINS_SERVICE_ACCOUNT: {
    _type: 'kube_namespace_contains_service_account',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.NAMESPACE._type,
    targetType: Entities.SERVICE_ACCOUNT._type,
  },
  NAMESPACE_CONTAINS_ROLE: {
    _type: 'kube_namespace_contains_role',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.NAMESPACE._type,
    targetType: Entities.ROLE._type,
  },
  NAMESPACE_CONTAINS_ROLE_BINDING: {
    _type: 'kube_namespace_contains_role_binding',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.NAMESPACE._type,
    targetType: Entities.ROLE_BINDING._type,
  },
  DEPLOYMENT_MANAGES_REPLICASET: {
    _type: 'kube_deployment_manages_replica_set',
    _class: RelationshipClass.MANAGES,
    sourceType: Entities.DEPLOYMENT._type,
    targetType: Entities.REPLICASET._type,
  },
  DEPLOYMENT_USES_CONTAINER_SPEC: {
    _type: 'kube_deployment_uses_container_spec',
    _class: RelationshipClass.USES,
    sourceType: Entities.DEPLOYMENT._type,
    targetType: Entities.CONTAINER_SPEC._type,
  },
  CONTAINER_SPEC_USES_VOLUME: {
    _type: 'kube_container_spec_uses_volume',
    _class: RelationshipClass.USES,
    sourceType: Entities.CONTAINER_SPEC._type,
    targetType: Entities.VOLUME._type,
  },
  CRONJOB_MANAGES_JOB: {
    _type: 'kube_cron_job_manages_job',
    _class: RelationshipClass.MANAGES,
    sourceType: Entities.CRONJOB._type,
    targetType: Entities.JOB._type,
  },
  NODE_HAS_IMAGE: {
    _type: 'kube_node_has_image',
    _class: RelationshipClass.HAS,
    sourceType: Entities.NODE._type,
    targetType: Entities.IMAGE._type,
  },
  REPLICA_SET_USES_IMAGE: {
    _type: 'kube_replica_set_uses_image',
    _class: RelationshipClass.USES,
    sourceType: Entities.REPLICASET._type,
    targetType: Entities.IMAGE._type,
  },
  REPLICASET_MANAGES_POD: {
    _type: 'kube_replica_set_manages_pod',
    _class: RelationshipClass.MANAGES,
    sourceType: Entities.REPLICASET._type,
    targetType: Entities.POD._type,
  },
  STATEFULSET_MANAGES_POD: {
    _type: 'kube_stateful_set_manages_pod',
    _class: RelationshipClass.MANAGES,
    sourceType: Entities.STATEFULSET._type,
    targetType: Entities.POD._type,
  },
  JOB_MANAGES_POD: {
    _type: 'kube_job_manages_pod',
    _class: RelationshipClass.MANAGES,
    sourceType: Entities.JOB._type,
    targetType: Entities.POD._type,
  },
  NODE_HAS_POD: {
    _type: 'kube_node_has_pod',
    _class: RelationshipClass.HAS,
    sourceType: Entities.NODE._type,
    targetType: Entities.POD._type,
  },
  POD_CONTAINS_CONTAINER: {
    _type: 'kube_pod_contains_container',
    _class: RelationshipClass.CONTAINS,
    sourceType: Entities.POD._type,
    targetType: Entities.CONTAINER._type,
  },
};

export const MappedRelationships: Record<
  string,
  StepMappedRelationshipMetadata
> = {
  CLUSTER_IS_AKS_CLUSTER: {
    _type: 'kube_cluster_is_azure_kubernetes_cluster',
    sourceType: Entities.CLUSTER._type,
    targetType: Entities.AZURE_KUBERNETES_CLUSTER._type,
    _class: RelationshipClass.IS,
    direction: RelationshipDirection.FORWARD,
  },
  CLUSTER_IS_GKE_CLUSTER: {
    _type: 'kube_cluster_is_google_container_cluster',
    sourceType: Entities.CLUSTER._type,
    targetType: Entities.GOOGLE_KUBERNETES_CLUSTER._type,
    _class: RelationshipClass.IS,
    direction: RelationshipDirection.FORWARD,
  },
};
