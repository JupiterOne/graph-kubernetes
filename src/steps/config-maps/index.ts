import {
  createDirectRelationship,
  RelationshipClass,
} from '@jupiterone/integration-sdk-core';
import { IntegrationStepContext } from '../../config';
import { CoreClient } from '../../kubernetes/clients/core';
import { KubernetesIntegrationStep } from '../../types';
import { Entities, IntegrationSteps, Relationships } from '../constants';
import { createConfigMapEntity } from './converters';

export async function fetchConfigMaps(
  context: IntegrationStepContext,
): Promise<void> {
  const { instance, jobState } = context;
  const { config } = instance;

  const client = new CoreClient(config);

  await jobState.iterateEntities(
    {
      _type: Entities.NAMESPACE._type,
    },
    async (namespaceEntity) => {
      await client.iterateNamespacedConfigMaps(
        namespaceEntity.name as string,
        async (configMap) => {
          const configMapEntity = createConfigMapEntity(configMap);
          await jobState.addEntity(configMapEntity);

          await jobState.addRelationship(
            createDirectRelationship({
              _class: RelationshipClass.CONTAINS,
              from: namespaceEntity,
              to: configMapEntity,
            }),
          );
        },
      );
    },
  );
}

export const configMapsSteps: KubernetesIntegrationStep[] = [
  {
    id: IntegrationSteps.CONFIGMAPS,
    name: 'Fetch ConfigMaps',
    entities: [Entities.CONFIGMAP],
    relationships: [Relationships.NAMESPACE_CONTAINS_CONFIGMAP],
    dependsOn: [IntegrationSteps.NAMESPACES],
    executionHandler: fetchConfigMaps,
    permissions: [
      {
        apiGroups: [''],
        resources: ['configmaps'],
        verbs: ['list'],
      },
    ],
  },
];
