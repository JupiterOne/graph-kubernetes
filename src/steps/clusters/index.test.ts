import {
  buildClusterAksRelationships,
  buildClusterGkeRelationships,
  buildClusterResourcesRelationships,
  fetchClusterDetails,
} from '.';
import { createDataCollectionTest } from '../../../test/recording';
import { integrationConfig } from '../../../test/config';
import { Entities, MappedRelationships, Relationships } from '../constants';
import { fetchNamespaces } from '../namespaces';
import { fetchConfigMaps } from '../config-maps';
import { RelationshipClass } from '@jupiterone/integration-sdk-core';

describe('#fetchClusterDetails', () => {
  test('should collect data', async () => {
    await createDataCollectionTest({
      recordingName: 'fetchClusterDetails',
      recordingDirectory: __dirname,
      integrationConfig,
      stepFunctions: [fetchClusterDetails],
      entitySchemaMatchers: [
        {
          _type: Entities.CLUSTER._type,
          matcher: {
            _class: ['Cluster'],
            schema: {
              additionalProperties: false,
              properties: {
                _type: { const: 'kube_cluster' },
                _rawData: {
                  type: 'array',
                  items: { type: 'object' },
                },
                name: { type: 'string' },
                displayName: { type: 'string' },
                server: { type: 'string' },
                skipTlsVerify: { type: 'boolean' },
              },
            },
          },
        },
      ],
    });
  });
});

describe('#buildClusterResourcesRelationships', () => {
  test('should collect data', async () => {
    await createDataCollectionTest({
      recordingName: 'buildClusterResourcesRelationships',
      recordingDirectory: __dirname,
      integrationConfig,
      stepFunctions: [
        fetchClusterDetails,
        fetchNamespaces,
        buildClusterResourcesRelationships,
      ],
      entitySchemaMatchers: [],
      relationshipSchemaMatchers: [
        {
          _type: Relationships.CLUSTER_CONTAINS_NAMESPACE._type,
          matcher: {
            schema: {
              properties: {
                _class: { const: RelationshipClass.CONTAINS },
                _type: { const: 'kube_cluster_contains_namespace' },
              },
            },
          },
        },
      ],
    });
  });
});

describe('#buildClusterAksRelationships', () => {
  test('should collect data', async () => {
    await createDataCollectionTest({
      recordingName: 'buildClusterAksRelationships',
      recordingDirectory: __dirname,
      integrationConfig,
      stepFunctions: [
        fetchClusterDetails,
        fetchConfigMaps,
        buildClusterAksRelationships,
      ],
      entitySchemaMatchers: [],
      relationshipSchemaMatchers: [
        {
          _type: MappedRelationships.CLUSTER_IS_AKS_CLUSTER._type,
          matcher: {
            schema: {
              properties: {
                _class: { const: RelationshipClass.IS },
                _type: { const: 'kube_cluster_is_azure_kubernetes_cluster' },
              },
            },
          },
        },
      ],
    });
  });
});

describe('#buildClusterGkeRelationships', () => {
  test('should collect data', async () => {
    await createDataCollectionTest({
      recordingName: 'buildClusterGkeRelationships',
      recordingDirectory: __dirname,
      integrationConfig,
      stepFunctions: [fetchConfigMaps, buildClusterGkeRelationships],
      entitySchemaMatchers: [],
      relationshipSchemaMatchers: [
        {
          _type: MappedRelationships.CLUSTER_IS_GKE_CLUSTER._type,
          matcher: {
            schema: {
              properties: {
                _class: { const: RelationshipClass.IS },
                _type: { const: 'kube_cluster_is_google_container_cluster' },
              },
            },
          },
        },
      ],
      options: {
        recordFailedRequests: true,
      },
    });
  });
});
