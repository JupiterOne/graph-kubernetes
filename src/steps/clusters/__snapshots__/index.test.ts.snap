// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#buildClusterAksRelationships should collect data: jobState 1`] = `
{
  "collectedEntities": [
    {
      "_class": [
        "Cluster",
      ],
      "_key": "kube_cluster:docker-desktop",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "name": "docker-desktop",
            "server": "https://192.168.49.2:8443",
            "skipTLSVerify": false,
            "tlsServerName": undefined,
          },
        },
      ],
      "_type": "kube_cluster",
      "createdOn": undefined,
      "displayName": "docker-desktop",
      "name": "docker-desktop",
      "server": "https://192.168.49.2:8443",
      "skipTlsVerify": false,
    },
  ],
  "collectedRelationships": [],
  "encounteredTypes": [
    "kube_cluster",
  ],
  "numCollectedEntities": 1,
  "numCollectedRelationships": 0,
}
`;

exports[`#buildClusterGkeRelationships should collect data: jobState 1`] = `
{
  "collectedEntities": [],
  "collectedRelationships": [],
  "encounteredTypes": [],
  "numCollectedEntities": 0,
  "numCollectedRelationships": 0,
}
`;

exports[`#buildClusterResourcesRelationships should collect data: jobState 1`] = `
{
  "collectedEntities": [
    {
      "_class": [
        "Cluster",
      ],
      "_key": "kube_cluster:docker-desktop",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "name": "docker-desktop",
            "server": "https://192.168.49.2:8443",
            "skipTLSVerify": false,
            "tlsServerName": undefined,
          },
        },
      ],
      "_type": "kube_cluster",
      "createdOn": undefined,
      "displayName": "docker-desktop",
      "name": "docker-desktop",
      "server": "https://192.168.49.2:8443",
      "skipTlsVerify": false,
    },
    {
      "_class": [
        "Group",
      ],
      "_key": "868d4703-b338-4051-98e4-b65b4f0a0006",
      "_rawData": [
        {
          "name": "default",
          "rawData": V1Namespace {
            "apiVersion": undefined,
            "kind": undefined,
            "metadata": V1ObjectMeta {
              "annotations": undefined,
              "creationTimestamp": 2024-01-11T15:50:15.000Z,
              "deletionGracePeriodSeconds": undefined,
              "deletionTimestamp": undefined,
              "finalizers": undefined,
              "generateName": undefined,
              "generation": undefined,
              "labels": {
                "kubernetes.io/metadata.name": "default",
              },
              "managedFields": [
                V1ManagedFieldsEntry {
                  "apiVersion": "v1",
                  "fieldsType": "FieldsV1",
                  "fieldsV1": {
                    "f:metadata": {
                      "f:labels": {
                        ".": {},
                        "f:kubernetes.io/metadata.name": {},
                      },
                    },
                  },
                  "manager": "kube-apiserver",
                  "operation": "Update",
                  "subresource": undefined,
                  "time": 2024-01-11T15:50:15.000Z,
                },
              ],
              "name": "default",
              "namespace": undefined,
              "ownerReferences": undefined,
              "resourceVersion": "39",
              "selfLink": undefined,
              "uid": "868d4703-b338-4051-98e4-b65b4f0a0006",
            },
            "spec": V1NamespaceSpec {
              "finalizers": [
                "kubernetes",
              ],
            },
            "status": V1NamespaceStatus {
              "conditions": undefined,
              "phase": "Active",
            },
          },
        },
      ],
      "_type": "kube_namespace",
      "active": undefined,
      "createdOn": 1704988215000,
      "deletionGracePeriodSeconds": undefined,
      "displayName": "default",
      "finalizers": [
        "kubernetes",
      ],
      "generation": undefined,
      "name": "default",
      "resourceVersion": "39",
      "status.phase": "Active",
    },
    {
      "_class": [
        "Group",
      ],
      "_key": "52615a92-75c3-4c4b-8b72-51be82a4bc21",
      "_rawData": [
        {
          "name": "default",
          "rawData": V1Namespace {
            "apiVersion": undefined,
            "kind": undefined,
            "metadata": V1ObjectMeta {
              "annotations": undefined,
              "creationTimestamp": 2024-01-11T15:50:15.000Z,
              "deletionGracePeriodSeconds": undefined,
              "deletionTimestamp": undefined,
              "finalizers": undefined,
              "generateName": undefined,
              "generation": undefined,
              "labels": {
                "kubernetes.io/metadata.name": "kube-node-lease",
              },
              "managedFields": [
                V1ManagedFieldsEntry {
                  "apiVersion": "v1",
                  "fieldsType": "FieldsV1",
                  "fieldsV1": {
                    "f:metadata": {
                      "f:labels": {
                        ".": {},
                        "f:kubernetes.io/metadata.name": {},
                      },
                    },
                  },
                  "manager": "kube-apiserver",
                  "operation": "Update",
                  "subresource": undefined,
                  "time": 2024-01-11T15:50:15.000Z,
                },
              ],
              "name": "kube-node-lease",
              "namespace": undefined,
              "ownerReferences": undefined,
              "resourceVersion": "35",
              "selfLink": undefined,
              "uid": "52615a92-75c3-4c4b-8b72-51be82a4bc21",
            },
            "spec": V1NamespaceSpec {
              "finalizers": [
                "kubernetes",
              ],
            },
            "status": V1NamespaceStatus {
              "conditions": undefined,
              "phase": "Active",
            },
          },
        },
      ],
      "_type": "kube_namespace",
      "active": undefined,
      "createdOn": 1704988215000,
      "deletionGracePeriodSeconds": undefined,
      "displayName": "kube-node-lease",
      "finalizers": [
        "kubernetes",
      ],
      "generation": undefined,
      "name": "kube-node-lease",
      "resourceVersion": "35",
      "status.phase": "Active",
    },
    {
      "_class": [
        "Group",
      ],
      "_key": "7a2ee4f7-12fd-494b-93f0-c639b1e63a72",
      "_rawData": [
        {
          "name": "default",
          "rawData": V1Namespace {
            "apiVersion": undefined,
            "kind": undefined,
            "metadata": V1ObjectMeta {
              "annotations": undefined,
              "creationTimestamp": 2024-01-11T15:50:15.000Z,
              "deletionGracePeriodSeconds": undefined,
              "deletionTimestamp": undefined,
              "finalizers": undefined,
              "generateName": undefined,
              "generation": undefined,
              "labels": {
                "kubernetes.io/metadata.name": "kube-public",
              },
              "managedFields": [
                V1ManagedFieldsEntry {
                  "apiVersion": "v1",
                  "fieldsType": "FieldsV1",
                  "fieldsV1": {
                    "f:metadata": {
                      "f:labels": {
                        ".": {},
                        "f:kubernetes.io/metadata.name": {},
                      },
                    },
                  },
                  "manager": "kube-apiserver",
                  "operation": "Update",
                  "subresource": undefined,
                  "time": 2024-01-11T15:50:15.000Z,
                },
              ],
              "name": "kube-public",
              "namespace": undefined,
              "ownerReferences": undefined,
              "resourceVersion": "20",
              "selfLink": undefined,
              "uid": "7a2ee4f7-12fd-494b-93f0-c639b1e63a72",
            },
            "spec": V1NamespaceSpec {
              "finalizers": [
                "kubernetes",
              ],
            },
            "status": V1NamespaceStatus {
              "conditions": undefined,
              "phase": "Active",
            },
          },
        },
      ],
      "_type": "kube_namespace",
      "active": undefined,
      "createdOn": 1704988215000,
      "deletionGracePeriodSeconds": undefined,
      "displayName": "kube-public",
      "finalizers": [
        "kubernetes",
      ],
      "generation": undefined,
      "name": "kube-public",
      "resourceVersion": "20",
      "status.phase": "Active",
    },
    {
      "_class": [
        "Group",
      ],
      "_key": "ce86116b-863b-4e74-979e-3c031a9aeee1",
      "_rawData": [
        {
          "name": "default",
          "rawData": V1Namespace {
            "apiVersion": undefined,
            "kind": undefined,
            "metadata": V1ObjectMeta {
              "annotations": undefined,
              "creationTimestamp": 2024-01-11T15:50:15.000Z,
              "deletionGracePeriodSeconds": undefined,
              "deletionTimestamp": undefined,
              "finalizers": undefined,
              "generateName": undefined,
              "generation": undefined,
              "labels": {
                "kubernetes.io/metadata.name": "kube-system",
              },
              "managedFields": [
                V1ManagedFieldsEntry {
                  "apiVersion": "v1",
                  "fieldsType": "FieldsV1",
                  "fieldsV1": {
                    "f:metadata": {
                      "f:labels": {
                        ".": {},
                        "f:kubernetes.io/metadata.name": {},
                      },
                    },
                  },
                  "manager": "kube-apiserver",
                  "operation": "Update",
                  "subresource": undefined,
                  "time": 2024-01-11T15:50:15.000Z,
                },
              ],
              "name": "kube-system",
              "namespace": undefined,
              "ownerReferences": undefined,
              "resourceVersion": "12",
              "selfLink": undefined,
              "uid": "ce86116b-863b-4e74-979e-3c031a9aeee1",
            },
            "spec": V1NamespaceSpec {
              "finalizers": [
                "kubernetes",
              ],
            },
            "status": V1NamespaceStatus {
              "conditions": undefined,
              "phase": "Active",
            },
          },
        },
      ],
      "_type": "kube_namespace",
      "active": undefined,
      "createdOn": 1704988215000,
      "deletionGracePeriodSeconds": undefined,
      "displayName": "kube-system",
      "finalizers": [
        "kubernetes",
      ],
      "generation": undefined,
      "name": "kube-system",
      "resourceVersion": "12",
      "status.phase": "Active",
    },
  ],
  "collectedRelationships": [
    {
      "_class": "CONTAINS",
      "_fromEntityKey": "kube_cluster:docker-desktop",
      "_key": "kube_cluster:docker-desktop|contains|868d4703-b338-4051-98e4-b65b4f0a0006",
      "_toEntityKey": "868d4703-b338-4051-98e4-b65b4f0a0006",
      "_type": "kube_cluster_contains_namespace",
      "displayName": "CONTAINS",
    },
    {
      "_class": "CONTAINS",
      "_fromEntityKey": "kube_cluster:docker-desktop",
      "_key": "kube_cluster:docker-desktop|contains|52615a92-75c3-4c4b-8b72-51be82a4bc21",
      "_toEntityKey": "52615a92-75c3-4c4b-8b72-51be82a4bc21",
      "_type": "kube_cluster_contains_namespace",
      "displayName": "CONTAINS",
    },
    {
      "_class": "CONTAINS",
      "_fromEntityKey": "kube_cluster:docker-desktop",
      "_key": "kube_cluster:docker-desktop|contains|7a2ee4f7-12fd-494b-93f0-c639b1e63a72",
      "_toEntityKey": "7a2ee4f7-12fd-494b-93f0-c639b1e63a72",
      "_type": "kube_cluster_contains_namespace",
      "displayName": "CONTAINS",
    },
    {
      "_class": "CONTAINS",
      "_fromEntityKey": "kube_cluster:docker-desktop",
      "_key": "kube_cluster:docker-desktop|contains|ce86116b-863b-4e74-979e-3c031a9aeee1",
      "_toEntityKey": "ce86116b-863b-4e74-979e-3c031a9aeee1",
      "_type": "kube_cluster_contains_namespace",
      "displayName": "CONTAINS",
    },
  ],
  "encounteredTypes": [
    "kube_cluster",
    "kube_namespace",
    "kube_cluster_contains_namespace",
  ],
  "numCollectedEntities": 5,
  "numCollectedRelationships": 4,
}
`;

exports[`#fetchClusterDetails should collect data: jobState 1`] = `
{
  "collectedEntities": [
    {
      "_class": [
        "Cluster",
      ],
      "_key": "kube_cluster:docker-desktop",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "name": "docker-desktop",
            "server": "https://192.168.49.2:8443",
            "skipTLSVerify": false,
            "tlsServerName": undefined,
          },
        },
      ],
      "_type": "kube_cluster",
      "createdOn": undefined,
      "displayName": "docker-desktop",
      "name": "docker-desktop",
      "server": "https://192.168.49.2:8443",
      "skipTlsVerify": false,
    },
  ],
  "collectedRelationships": [],
  "encounteredTypes": [
    "kube_cluster",
  ],
  "numCollectedEntities": 1,
  "numCollectedRelationships": 0,
}
`;
